### PR TITLE
Mam remapping opt

### PR DESCRIPTION
**Pull request by grendello**
Fixes: https://github.com/xamarin/xamarin-android/issues/7020

Optimize MAM remapping by generating native data at build time, 
to be included in `libxamarin-app.so` and eliminating the need
to parse XML and use managed dictionaries at run time when
remapping types/methods.

### CODE REVIEW

File: src/Mono.Android/Android.Runtime/AndroidRuntime.cs [hunk 1]
```diff
@@ -317,123 +324,44 @@ protected override IEnumerable<string> GetSimpleReferences (Type type)
 			};
 		}
 
+		[DllImport (AndroidRuntime.InternalDllName, CallingConvention = CallingConvention.Cdecl)]
+		static extern string? _monodroid_lookup_replacement_type (string jniSimpleReference);
```
Comment (jonpryor) [inline src/Mono.Android/Android.Runtime/AndroidRuntime.cs:328]: Unless `_monodroid_lookup_replacement_type()` is allocating memory with the "COM task memory allocator" (iirc `g_malloc()` & `g_free()` on Mono + Unixy platforms), this is Not Good™.   This should probably be `IntPtr`.

Comment (jonpryor) [reply src/Mono.Android/Android.Runtime/AndroidRuntime.cs:328]: See also:

* https://www.mono-project.com/docs/advanced/pinvoke/#strings-as-return-values
* https://www.mono-project.com/docs/advanced/pinvoke/#classes-and-structures-as-return-values

Comment (grendello) [reply src/Mono.Android/Android.Runtime/AndroidRuntime.cs:328]: > Unless `_monodroid_lookup_replacement_type()` is allocating memory with the "COM task memory allocator" (iirc `g_malloc()` & `g_free()` on Mono + Unixy platforms), this is Not Good™. This should probably be `IntPtr`.

Memory is in a read-only string section in ELF

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs [hunk 2]
```diff
@@ -45,7 +45,7 @@ public sealed class ApplicationConfig
 			public bool   uses_assembly_preload;
 			public bool   is_a_bundled_app;
 			public bool   broken_exception_transitions;
-			public bool   instant_run_enabled;
+			public bool   instant_run_enabled ;
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/EnvironmentHelper.cs:48]:
```suggestion
			public bool   instant_run_enabled;
```

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Resources/Base/BuildReleaseArm64SimpleDotNet.apkdesc [hunk 3]
```diff
@@ -116,5 +83,5 @@
       "Size": 1904
     }
   },
-  "PackageSize": 3492724
+  "PackageSize": 3012575
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Resources/Base/BuildReleaseArm64SimpleDotNet.apkdesc:86]: Yay!

File: src/Mono.Android/Android.Runtime/AndroidRuntime.cs [hunk 4]
```diff
@@ -243,6 +243,13 @@ public override void DeleteWeakGlobalReference (ref JniObjectReference value)
 	}
 
 	class AndroidTypeManager : JniRuntime.JniTypeManager {
+		struct MAMReplacementMethod
```
[the remark this answers is not in the document]
Comment (grendello) [reply src/Mono.Android/Android.Runtime/AndroidRuntime.cs:246]: It's a very good point, perhaps we should prefix the types with something more generic (I do want to have a prefix that identifies the structures as belonging to a particular feature). `MAM` -> `Remapping`, perhaps? @jonpryor?

[the remark this answers is not in the document]
Comment (jonpryor) [reply src/Mono.Android/Android.Runtime/AndroidRuntime.cs:246]: @grendello: yes, let's go with `JniRemapping` or something as a prefix.

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 5]
```diff
@@ -1620,6 +1624,41 @@ because xbuild doesn't support framework reference assemblies.
...
+        Inputs="$(MSBuildProjectFullPath)"
+        Outputs="@(_AndroidRemapAssemblySource)">
+  <GenerateJniRemappingNativeCode
+      OutputDirectory="$(_NativeAssemblySourceDir)"
+      GenerateEmptyCode="True"
+      SupportedAbis="@(_BuildTargetAbis)"
+   />
+</Target>
+
+<Target Name="_GenerateAndroidRemapNativeCode"
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1650]: Are we expecting these targets to run on every build? i.e no support for skipping if nothing changes between builds?

Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:1650]: We require the sources they generate for all builds, even if the remapping data isn't there (libxamarin-app.so ABI requirements). However, if they already generated the LLVM IR code and there are no changes, then they don't need to regenerate the sources.

**Comment by jonpryor** [comment]
```markdown
[Xamarin.Android.Build.Tasks, monodroid] Optimize member remapping (#7059)

Fixes: https://github.com/xamarin/xamarin-android/issues/7020

Context: f6f11a5a797cd8602854942dccb0f2b1db57c473

Commit f6f11a5a introduced type & member remapping.  The problem with
its approach was that it used XML as the representation format, which
needed to be parsed during process startup.  This would contribute to
app startup slowdowns if there were any remappings, but even if there
weren't any remappings, the minimum App size increased by ~490KB, due
to the added dependencies on `System.Private.Xml.dll` & more.

This app size increase is not ideal.

Remove most of the `.apk` size increases by moving the remapping info
into `libxamarin-app.so`.  If no remappings are used, then the size
increase to `libxamarin-app.so` is negligible, and the `.apk` size is
only 61KB larger than pre-f6f11a5a.


~~ API Changes ~~

`Android.Runtime.AndroidTypeManager` gains the following members:

	partial class AndroidTypeManager {
	    struct JniRemappingReplacementMethod {
	        public string target_type, target_name;
	        public bool   is_static;
	    }
	    static extern byte* _monodroid_lookup_replacement_type (
	            string jniSimpleReference
	    );
	    static extern JniRemappingReplacementMethod* _monodroid_lookup_replacement_method_info (
	            string jniSourceType,
	            string jniMethodName,
	            string jniMethodSignature
	    );
	}

`AndroidTypeManager._monodroid_lookup_replacement_type()` replaces
the `JNIEnv.ReplacementTypes` dictionary from f6f11a5a.

`AndroidTypeManager._monodroid_lookup_replacement_method_info()`
replaces the `JNIEnv.ReplacementMethods` dictionary from f6f11a5a.

Both `_monodroid_lookup_replacement_type()` and
`_monodroid_lookup_replacement_method_info()` are P/Invokes into
`libxamarin-app.so`.


~~ `libxamarin-app.so` Changes ~~

The contents of the `@(_AndroidRemapMembers)` item group are now
stored within `libxamarin-app.so`, with the following structure:

	const uint32_t                          jni_remapping_replacement_method_index_entry_count;
	const JniRemappingIndexTypeEntry        jni_remapping_method_replacement_index[];

	const uint32_t                          jni_remapping_replacement_type_count;
	const JniRemappingTypeReplacementEntry  jni_remapping_type_replacements[];

	struct JniRemappingString {
	    const uint32_t                  length;
	    const char                     *str;
	};
	struct JniRemappingReplacementMethod {
	    const char                     *target_type, *target_name;
	    const bool                      is_static;
	};
	struct JniRemappingIndexMethodEntry {
	    JniRemappingString              name, signature;
	    JniRemappingReplacementMethod   replacement
	};
	struct struct JniRemappingIndexTypeEntry {
	    JniRemappingString              name;
	    uint32_t                        method_count;
	    JniRemappingIndexMethodEntry   *methods;
	};
	struct JniRemappingTypeReplacementEntry {
	    JniRemappingString              name;
	    const char                     *replacement;
	};
	const char *
	_monodroid_lookup_replacement_type (const char *);

	const JniRemappingReplacementMethod*
	_monodroid_lookup_replacement_method_info (const char *jniSourceType, const char *jniMethodName, const char *jniMethodSignature);

Referring to the `<replacements/>` XML from f6f11a5a in
`@(_AndroidRemapMembers)`:

  * `//replace-type/@from` fills `JniRemappingTypeReplacementEntry::name`,
    `//replace-type/@to` fills `JniRemappingTypeReplacementEntry::replacement`,
    and `_monodroid_lookup_replacement_type()` performs a linear
    search over `jni_remapping_type_replacements`.

  * `//replace-method/@source-type` fills `JniRemappingIndexTypeEntry::name`,
    `//replace-method/@source-method-name` fills `JniRemappingIndexMethodEntry::name`,
    `//replace-method/@source-method-signature` fills `JniRemappingIndexMethodEntry::signature`,
    `//replace-method/@target-type` fills `JniRemappingReplacementMethod::target_type`,
    `//replace-method/@target-method-name` fills `JniRemappingReplacementMethod::target_name`,
    `//replace-method/@target-method-signature` and
    `//replace-method/@target-method-parameter-count` are *ignored*,
    `//replace-method/@target-method-instance-to-static` fills `JniRemappingReplacementMethod::is_static`,
    and `_monodroid_lookup_replacement_method_info()` performs a
    search over `jni_remapping_method_replacement_index` looking for
    entries with "matching" type names, method names, and method
    signatures, and once a match is found it returns a pointer to the
    `JniRemappingReplacementMethod` instance.

Co-authored-by: Jonathan Peppers <jonathan.peppers@gmail.com>
```

**Pull request merged**